### PR TITLE
HIVE-29703: LLAP: Parquet footer larger than alloc.max throws RuntimeException during caching

### PR DESCRIPTION
**Pull request by deniskuzZ**
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
  - Allocate the remainder buffer with smallSize, not length.                                                                                                                                                    
  - Use maxAlloc <= length so a footer of exactly maxAlloc is held in one full buffer.                                                                                                                                                                                                                            

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
When LLAP caches a Parquet file footer whose size exceeds hive.llap.io.allocator.alloc.max, the query fails with:     
````                                                                                         
  java.lang.RuntimeException: Trying to allocate 35405479; max is 33554432
````

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
````
mvn test -pl llap-server -Dtest='TestOrcMetadataCache#testStreamFooterEqualToMaxAlloc+testStreamFooterLargerThanMaxAlloc
````

### CODE REVIEW

File: llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java [hunk 1]
```diff
@@ -257,39 +257,43 @@ public LlapBufferOrBuffers putFileMetadata(Object fileKey, int length, InputStre
   }
 
 
-  @SuppressWarnings({ "rawtypes", "unchecked" })
+  @SuppressWarnings("unchecked")
   private LlapBufferOrBuffers wrapBbForFile(LlapBufferOrBuffers result,
```
Comment (Aggarwal-Raghav) [inline llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java:261]: Question? If there are any exceptions like `IOException` or `AllocatorOutOfMemoryException` in `readIntoCacheBuffer()`/` allocateMultiple()` shouldn't there be a `finally` block to  deallocate  the previously allocated buffer memory? 

In `readIntoCacheBuffer()` there is no finally block. In `allocateMultiple()` there is finally block but will deallocate the current buffer not previously allocated buffer.

Comment (deniskuzZ) [reply llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java:261]: need to check, but that's seems to be a diff issue

Comment (Aggarwal-Raghav) [reply llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java:261]: yeah, its outside the scope of current JIRA. If it's ok , can I pick it up as separate JIRA?  Will come up with a UT for this?

Comment (deniskuzZ) [reply llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java:261]: oh, haven't noticed your latest comment and decided to fix it here: 3bf78175de8f83c6d9f255ae4ffda546a04273a7

Comment (Aggarwal-Raghav) [reply llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java:261]: No worries, resolution of issue is important.

**Comment by Aggarwal-Raghav** [comment]
LGTM +1

**Review by Aggarwal-Raghav — APPROVED** [review]

File: llap-server/src/test/org/apache/hadoop/hive/llap/cache/TestMetadataCache.java [hunk 2]
```diff
@@ -185,6 +185,104 @@ public void testBuffers() throws Exception {
...
+    MetadataCache cache = new MetadataCache(alloc, mm, new DummyCachePolicy(), true, metrics);
+
+    Object fileKey = new Object();
+    // One byte short of the footer, so reading the last chunk hits EOF part-way through.
+    ByteArrayInputStream truncated = new ByteArrayInputStream(new byte[length - 1]);
+    try {
+      cache.putFileMetadata(fileKey, length, truncated, null, null);
+      fail("Expected an EOFException for a truncated footer stream of length " + length);
+    } catch (IOException expected) {
+      // expected
```
Comment (abstractdog) [inline llap-server/src/test/org/apache/hadoop/hive/llap/cache/TestMetadataCache.java:237]: assert the exact EOF here that we're expecting

File: llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java [hunk 3]
```diff
@@ -257,41 +257,85 @@ public LlapBufferOrBuffers putFileMetadata(Object fileKey, int length, InputStre
...
-      allocator.allocateMultiple(smallBuffer, length, null, isStopped);
-      readIntoCacheBuffer(stream, smallSize, smallBuffer[0]);
-      if (largeBuffers == null) {
-        return smallBuffer[0]; // This is the overwhelmingly common case.
-      } else {
-        LlapMetadataBuffer<Object>[] cacheData = new LlapMetadataBuffer[largeBuffers.length + 1];
-        System.arraycopy(largeBuffers, 0, cacheData, 0, largeBuffers.length);
-        cacheData[largeBuffers.length] = smallBuffer[0];
-        return new LlapMetadataBuffers<>(cacheData);
+      int smallSize = length % maxAlloc;
```
Comment (abstractdog) [inline llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java:299]: I believe this is rather "remainderSize" for clarity

File: llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java [hunk 4]
```diff
@@ -257,41 +257,85 @@ public LlapBufferOrBuffers putFileMetadata(Object fileKey, int length, InputStre
...
+    }
+    // Larger footers are split across maxAlloc-sized chunks, the last one holding the remainder.
+    LlapMetadataBuffer<Object>[] largeBuffers = new LlapMetadataBuffer[length / maxAlloc];
+    for (int i = 0; i < largeBuffers.length; ++i) {
+      largeBuffers[i] = new LlapMetadataBuffer<>(fileKey, tag);
+    }
+    // allocateMultiple is all-or-nothing: on success every chunk is allocated; on failure it
+    // releases whatever it reserved, so a throw here needs no cleanup from us.
+    allocator.allocateMultiple(largeBuffers, maxAlloc, null, isStopped);
+    LlapMetadataBuffer<Object> smallBuffer = null;
```
Comment (abstractdog) [inline llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java:293]: `remainderBuffer` to clarify what is "small" for

File: llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java [hunk 5]
```diff
@@ -257,41 +257,85 @@ public LlapBufferOrBuffers putFileMetadata(Object fileKey, int length, InputStre
...
-      allocator.allocateMultiple(largeBuffers, maxAlloc, null, isStopped);
+    }
+    // Larger footers are split across maxAlloc-sized chunks, the last one holding the remainder.
+    LlapMetadataBuffer<Object>[] largeBuffers = new LlapMetadataBuffer[length / maxAlloc];
+    for (int i = 0; i < largeBuffers.length; ++i) {
+      largeBuffers[i] = new LlapMetadataBuffer<>(fileKey, tag);
+    }
+    // allocateMultiple is all-or-nothing: on success every chunk is allocated; on failure it
+    // releases whatever it reserved, so a throw here needs no cleanup from us.
+    allocator.allocateMultiple(largeBuffers, maxAlloc, null, isStopped);
```
Comment (abstractdog) [inline llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java:291]: nit, maybe add a line break here: next line (`smallBuffer = null`) is about to handle a different thing

**Review by abstractdog — CHANGES_REQUESTED** [review]
only some minor nits, ideas, rest LGTM

File: llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java (hunk 5 above)
Comment (deniskuzZ) [reply llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java:291]: added

File: llap-server/src/test/org/apache/hadoop/hive/llap/cache/TestMetadataCache.java (hunk 2 above)
Comment (deniskuzZ) [reply llap-server/src/test/org/apache/hadoop/hive/llap/cache/TestMetadataCache.java:237]: narrowed the catch to EOFException

File: llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java (hunk 3 above)
Comment (deniskuzZ) [reply llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java:299]: i would prefer to keep the original naming so we stay consistent with other var names in particular largeBuffers

File: llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java (hunk 4 above)
Comment (deniskuzZ) [reply llap-server/src/java/org/apache/hadoop/hive/llap/io/metadata/MetadataCache.java:293]: https://github.com/apache/hive/pull/6584#discussion_r3579169276

**Review by abstractdog — APPROVED** [review]
LGTM
